<?xml version="1.0"?>
<launch>
    <!-- TR200 모듈화된 센서 기반 안전 제어 시스템 런치 파일 -->
    
    <!-- 로봇 연결 파라미터 -->
    <arg name="robot_ip" default="169.254.128.2" />
    <arg name="robot_port" default="5480" />
    
    <!-- 센서 설정 파라미터 -->
    <arg name="min_obstacle_distance" default="0.5" />
    <arg name="warning_distance" default="0.8" />
    <arg name="safe_distance" default="1.0" />
    
    <!-- 제어 설정 파라미터 -->
    <arg name="normal_speed" default="0.2" />
    <arg name="slow_speed" default="0.1" />
    <arg name="stop_speed" default="0.0" />
    <arg name="control_frequency" default="20.0" />
    
    <!-- 토픽 설정 파라미터 -->
    <arg name="scan_topic" default="/scan" />
    <arg name="processed_scan_topic" default="/processed_scan" />
    <arg name="safe_cmd_vel_topic" default="/safe_cmd_vel" />
    <arg name="safety_status_topic" default="/safety_status" />
    <arg name="robot_status_topic" default="/robot_status" />
    
    <!-- YAML 파일에서 파라미터 로드 (현업 표준 방식) -->
    <rosparam file="$(find tr200_simple_control)/config/tr200_sensor_safety_params.yaml" />
    
    <!-- 모듈 1: 센서 데이터 처리 노드 -->
    <node 
        name="sensor_processor" 
        pkg="tr200_simple_control" 
        type="sensor_processor_node.py" 
        output="screen"
        respawn="true"
        respawn_delay="5">
        
        <!-- 로봇 연결 파라미터 -->
        <param name="robot_ip" value="$(arg robot_ip)" />
        <param name="robot_port" value="$(arg robot_port)" />
        
        <!-- 토픽 설정 -->
        <param name="scan_topic" value="$(arg scan_topic)" />
        <param name="processed_scan_topic" value="$(arg processed_scan_topic)" />
    </node>
    
    <!-- 모듈 2: 안전 제어 노드 -->
    <node 
        name="safety_controller" 
        pkg="tr200_simple_control" 
        type="safety_controller_node.py" 
        output="screen"
        respawn="true"
        respawn_delay="5">
        
        <!-- 센서 설정 파라미터 -->
        <param name="min_obstacle_distance" value="$(arg min_obstacle_distance)" />
        <param name="warning_distance" value="$(arg warning_distance)" />
        <param name="safe_distance" value="$(arg safe_distance)" />
        
        <!-- 제어 설정 파라미터 -->
        <param name="normal_speed" value="$(arg normal_speed)" />
        <param name="slow_speed" value="$(arg slow_speed)" />
        <param name="stop_speed" value="$(arg stop_speed)" />
        <param name="control_frequency" value="$(arg control_frequency)" />
        
        <!-- 토픽 설정 -->
        <param name="processed_scan_topic" value="$(arg processed_scan_topic)" />
        <param name="safe_cmd_vel_topic" value="$(arg safe_cmd_vel_topic)" />
        <param name="safety_status_topic" value="$(arg safety_status_topic)" />
    </node>
    
    <!-- 모듈 3: 로봇 제어 노드 -->
    <node 
        name="robot_driver" 
        pkg="tr200_simple_control" 
        type="robot_driver_node.py" 
        output="screen"
        respawn="true"
        respawn_delay="5">
        
        <!-- 로봇 연결 파라미터 -->
        <param name="robot_ip" value="$(arg robot_ip)" />
        <param name="robot_port" value="$(arg robot_port)" />
        
        <!-- 토픽 설정 -->
        <param name="safe_cmd_vel_topic" value="$(arg safe_cmd_vel_topic)" />
        <param name="robot_status_topic" value="$(arg robot_status_topic)" />
    </node>
    
    <!-- 선택적: RViz 시각화 -->
    <arg name="use_rviz" default="false" />
    <node if="$(arg use_rviz)" 
          name="rviz" 
          pkg="rviz" 
          type="rviz" 
          args="-d $(find tr200_simple_control)/config/tr200_sensor_safety.rviz" />
    
    <!-- 선택적: 외부 제어를 위한 키보드 텔레옵 -->
    <arg name="use_teleop" default="false" />
    <node if="$(arg use_teleop)" 
          name="keyboard_teleop" 
          pkg="teleop_twist_keyboard" 
          type="teleop_twist_keyboard.py" 
          output="screen">
        <remap from="cmd_vel" to="/external_cmd_vel" />
    </node>
    
    <!-- 선택적: 로그 레코딩 -->
    <arg name="record_logs" default="false" />
    <node if="$(arg record_logs)" 
          name="rosbag_record" 
          pkg="rosbag" 
          type="record" 
          args="-a -o $(find tr200_simple_control)/logs/tr200_modular_system" />
    
</launch>
